### DISCUSSION TITLE
Increase transparency of the development-process and -state of F#

### DISCUSSION THREAD

**Discussion by SchlenkR** [fsharp/fslang-design#749] [General]
As briefly discussed on https://twitter.com/dawe70704856/status/1696430171882975604, my personal feeling is that transparency in the development of F# has decreased lately. Maybe that feeling is wrong, but I can give some examples representative for it that I hope might clarify what this is initially about. My very general expectation of a major open source project which is open for partifipation and contribution regarding transparency is:

* it should be possible to understand on what is currently worked (maybe also by whom),
* what can be expected to change / added in the near future,
* what is the approximate direction on a longer-term view.

Surely there are other points and you don't have to share the points mentioned. That F# is not a pure community project, but a largely Microsoft funded and directed investment (thanks for that; no irony here), is clear to me. I'm also not concerned with democratizing decision-making processes here. Even the way decisions come about doesn't necessarily have to be open or well-reasoned (to me). What I think is necessary, however, is "what" is happening or will happen. Why's that important, actually? A subjective answer: for me it was and is always a great experience to get in touch with people from "the community". F# is not a huge thing, but I always had the impression that for many, a very specific feeling is of importance, which goes beyond the (in every respect remarkable) language F#. To have the feeling of "something is moving and I can make a difference": C#, TypeScript - they don't need that. F#, perhaps, does. For me it was always of great importance; I liked it and could write a lot about it! I simply connect a bunch of good feelings and memories with F#; as crazy as it sounds: That's the truth.

From the README of this repo:

> Much of the work happens by contributions via RFCs and most features now proceed from initial approval all the way to implementation through community and enterprise contributions. The planning and progress process for features is intended to be transparent and participative.

So transparency and participation are intended - I had no doubt about that. So it's more about what it means or how it can be achieved. By the way: As a passive part of the F# language and tooling development I only refer to transparency - not to active participation here.

What has changed concretely from my point of view: I used to focus on fslang-suggestions and fslang-design. While the former gave (and still does) a rough sense of what might happen and why, fslang-design was the place to understand what will most likely happen or is currently happening. But also what has happened - so important as a de facto place of prose specification and explanation. That was enough for me to think I was "well informed". This didn't require F# Discord, Slack, Twitter or so. Meanwhile, I have the impression that important things no longer catch on. Maybe I simply do not know where to look - can that be?

Again: This is a very subjective account. If that opinion shared, I could imagine a few simple things that could lead to more transparency:

- Maintain fslang-design more actively as a mirror to things that are happening or planned.
- Involve people from the community to act as loudspeakers / amplifyers / multipliers through various channels (besides great shows like "Languages and Community Standup"). 

Bottom line, I just think it's sad when hard work (from everyone who participates) is so invisible. Thanks to everyone - internal and external, paid or unpaid, pro or amateur - for the great commitment, which also benefits developers like me.

Have a nice day.

**Comment by cartermp** [comment]
What in particular is invisible? Both features involved in the announcement are sourced from the fslang-suggestions repo:

https://github.com/fsharp/fslang-suggestions/issues/1262
https://github.com/fsharp/fslang-suggestions/issues/506

And the first is precisely a case of community involvement, wherein something originally closed as "won't do" was argued for openly and eventually accepted into the language.

**Reply by SchlenkR** [reply]
As far as I can see, they are not visible in the fslang-design repo. 

There it says:

All open RFCs that have not been released live under /RFCs.

All RFCs that have been implemented in a preview of F# live under /preview.

All release language and FSharp.Core RFCs live under a folder with their respective release.

All tooling RFCs (typically for cross-cutting, cross-editor tooling) live under /tooling.

It felt that it was not obvious that these features are currently being implemented. There are a lot of issues in fslang-suggestions.

Maybe my expectations were just wrong in some way - possible.

**Reply by smoothdeveloper** [reply]
@RonaldSchlenker, curious if you have the dotnet/fsharp repository "watched"? I feel it is a good way to catch up with which PR is being worked on (after removing the dotnet-bot noise).

Most of the work is occurring in public PRs, and by watching the repository, I do receive emails related to the issues & PRs.

I'm a bit wary that by asking more of status update to occur, we'll lose some of the energy which goes into the ground work.

There are few status update issues that provides some heads up on next efforts popping up once in a while, there is also https://github.com/orgs/dotnet/projects/126/views/33?filterQuery=is%3Aopen+&sliceBy%5BcolumnId%5D=Assignees which allows to see who is assigned to what.

In this respect, I feel it provides enough transparency.

**Reply by cartermp** [reply]
Yeah, all work is done in the open. Is this a complaint that a particular documented process isn't being followed? That I can see. But I don't see how things are lacking transparency. You're either engaged in the process or you're not.

**Reply by vzarytovskii** [reply]
As for tooling RFCs and changes, we don't require those for 99.9% of cases, as they make little sense.

As was mentioned before - everything planned can be followed in the project and this repo, as for RFCs, there's no process of moving them systematically. 

Perhaps there should be, not sure for now how it will be enforced. 

We don't use any internal trackers or planning tools.

**Reply by SchlenkR** [reply]
To clarify, there is also no demand or similar - it was thought as feedback or idea. Maybe this repo wasn't a good place to start this kind of discussion, but ok - here we are :)

> Is this a complaint that a particular documented process isn't being followed?

No, it's not a complaint, but an observation. Here's an analogy: 

Every morning, you go to a Kiosk and buy the newspaper of your choice, as a serious and reliable source of information. The process is convenient, well-known and affordable. One day, you recognize: Kiosk is there; newspaper not. Next day, same issue, and so forth. You ask yourself: "How to substitute all that?". This repo was the go-to place for getting an impression of the points mentioned above. It seems to me that it's not the case anymore. If that's not the case anymore, it's ok for me. I asked myself (and others) where to find the information that I always found here. TBH, I don't watch dotnet/fsharp, but that might be a good source, so I'll do that, thanks @smoothdeveloper .

> But I don't see how things are lacking transparency.

Let me put it a little differently: Important steps and contributions don't seem to be very prominent. Again, an observation and my personal impression. I'll have a more frequent look at dotnet/fsharp.

**Reply by SchlenkR** [reply]
>We don't use any internal trackers or planning tools.

❤️

**Reply by krauthaufen** [reply]
I also find it difficult to see what's currently happening with F#, since there are a lot of RFCs (some pretty old) and almost all the updates I get are about formatting/style. I second the idea of some landing-page showing what is currently being worked on. Maybe I'm just missing something.

**Reply by vzarytovskii** [reply]
> I also find it difficult to see what's currently happening with F#, since there are a lot of RFCs (some pretty old) and almost all the updates I get are about formatting/style. I second the idea of some landing-page showing what is currently being worked on. Maybe I'm just missing something.

This is pretty much just issues and PRs in this repo. It shows in detail what's happening here. Project is also good source of what's team is planning (though it's oftentimes randomized with current PR reviews, bugfixes and some additional work, which is reflected in this repo activity too). 

If more detailed info is desired, I personally will not be opposed to FSSF maintaining such separate page based on said sources.

**Reply by SchlenkR** [reply]
> If more detailed info is desired, I personally will not be opposed to FSSF maintaining such separate page based on said sources.

That sounds interesting @vzarytovskii, and I had something similar in mind, too. Could be a blog with small summaries / currated links of the current state of F# language / core / tooling progression; with a narrow focus on exactly these things. Could be done via FSSF, but also a community thing in addition to Sergey's F# Weekly (which is my go-to place and source for general F# and related infos), and Amplifying F# (as a hands-on support for getting-things-done in the compiler and ecosystem).

The goal(s) of such a thing could be

* a motivation for people willing to contribute; make their work prominent in the F# community, and hopefully to externals
* one more small piece that helps creating the impression of an "F# is alive" feeling (those feelings - I think they are important for people and you don't always have them in view with all the other challenges).
* people who are not able to dig in deep into fslang-sugg/design or dotnet/fsharp get a summary of important F# lang progression
* arouse interest in people who have had less to do with F# so far
* etc.

Maybe I'll give it a try... Thanks for all your comments (more appreciated if you think it's a thing worth to try).

**Reply by T-Gro** [reply]
I understand the desire to have a middle-ground between waiting for an official "F# vNext just released" blog post and actively watching the dotnet/fsharp repo.

That being said, I also have to emphasize that chances and timelines for a PR in progress are not predictable at all. I have seen the wave of negative feedback for "Tentative C# 12 feature list" and was very disappointed by it. Arousing interest is good, frustration when items do not eventually make it - not so much.

If anyone wants to curate a list of finished and WIP pull requests that relate to a language change, I think I can offer marking them with a specific label whenever I come across them ("e.g. 'Area-Language-Change'). That way, a URL with specified GH filter can be used to irregularly watch them without having to go trough all the other PRs (bugfixes, bots, tooling changes,...).

**Reply by Happypig375** [reply]
@T-Gro Unlike C# language design, here in F# land if something is not on the plans, we can easily invite those dissatisfied to suggest, write RFC or even go for an implementation. C# design is heavily gated by the team from Microsoft while F# design is much more open and up-for-grabs. I think we will be in a better shape than C# here.

**Comment by Happypig375** [comment]
Follow-up change: https://github.com/dotnet/fsharp/issues/15901

**Comment by Smaug123** [comment]
It's not ideal, but these days there's a pretty canonical list of new language features at https://github.com/dotnet/fsharp/blob/0bfd490b47fd35c4a48c88a1b2fd8c50b724f3e1/src/Compiler/Facilities/LanguageFeatures.fs#L113 . The two new features mentioned above are `LanguageFeature.ConstraintIntersectionOnFlexibleTypes` and `LanguageFeature.AccessorFunctionShorthand`. It would be kind of nice to have some tooling autogenerating release notes for new language features!

**Reply by kerams** [reply]
I don't think generated release notes are a step in the right direction. They are sterile, and give you no context.

At the risk of sounding like a broken record, regular blog posts (once a month or two, and with every (preview) release of .NET and VS) showcasing what has been done (language- *and* tooling-wise) and is available to try out, what the priorities are, what RFCs have been approved, would reach a much larger audience. The downside is that someone needs to (consistently) dedicate time to this. I do, however, feel that it is important enough.

On a related note, it's also a shame that I don't remember the last time I saw any mention of F# in [VS release notes](https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview#17.8.0-pre.1.0).

**Reply by Smaug123** [reply]
> I don't think generated release notes are a step in the right direction. They are sterile, and give you no context.

Well, it's not like the *text* of these things would be autogenerated :P the no-thought implementation would be a map of LanguageFeature to blog-text Markdown string.

**Reply by SchlenkR** [reply]
> On a related note, it's also a shame that I don't remember the last time I saw any mention of F# in [VS release notes](https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview#17.8.0-pre.1.0).

Yes, that's sad. When I talk to my clients (mostly C# / TS / whatever-average-client-stuff), they don't even see F#, or have heared about it, but don't really understand what's going on or if it's still "alive". Part of the reality - like it or not - is this: The louder one shouts, the more often one does it, the more visible one becomes, where visibility is at least a factor in becoming or staying relevant. There's a proverb, that goes like this: "As a politician, it's better to have bad publicity than none". Now I'm not saying that it's good to spam people, shout at them or so - and it's not necessary - because there's so much good stuff here! Instead of having no or bad publicity, we can have good one! In the end, there might be less publicity than we could have. Personally, I'm not good at marketing, and I have no clue how to do it, and I love spending my free evenings developing software, rather than doing a marketing campaign or something. But in the end, I'm just convinced that F# is one of the most practical languages out there, and I'm happy to see that so many people put effort in it, with enthusiasm, and that's something that's unique - at least for the things I know. Hopefully, I'll find a way, time, mood and spirit for making an editorial format that will make everything that's going on in this great community more visible; at least a little bit. SOmething about the language progression, the community, the people. Failure is always an option; maybe I won't be able or it will just be crap but I will try...

**Reply by vzarytovskii** [reply]
> > On a related note, it's also a shame that I don't remember the last time I saw any mention of F# in [VS release notes](https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview#17.8.0-pre.1.0).
> 
> Yes, that's sad. When I talk to my clients (mostly C# / TS / whatever-average-client-stuff), they don't even see F#, or have heared about it, but don't really understand what's going on or if it's still "alive". Part of the reality - like it or not - is this: The louder one shouts, the more often one does it, the more visible one becomes, where visibility is at least a factor in becoming or staying relevant. There's a proverb, that goes like this: "As a politician, it's better to have bad publicity than none". Now I'm not saying that it's good to spam people, shout at them or so - and it's not necessary - because there's so much good stuff here! Instead of having no or bad publicity, we can have good one! In the end, there might be less publicity than we could have. Personally, I'm not good at marketing, and I have no clue how to do it, and I love spending my free evenings developing software, rather than doing a marketing campaign or something. But in the end, I'm just convinced that F# is one of the most practical languages out there, and I'm happy to see that so many people put effort in it, with enthusiasm, and that's something that's unique - at least for the things I know. Hopefully, I'll find a way, time, mood and spirit for making an editorial format that will make everything that's going on in this great community more visible; at least a little bit. SOmething about the language progression, the community, the people. Failure is always an option; maybe I won't be able or it will just be crap but I will try...

Main reason is that we don't have technical writers on the team, and people are usually busy with tasks they're working on, so release notes are not simply sent to editorial team.

**Reply by SchlenkR** [reply]
> Main reason is that we don't have technical writers on the team, and people are usually busy with tasks they're working on, so release notes are not simply sent to editorial team.

Yes, I can understand that absolutely. Devs are devs - not PR or marketing people. It's a very different role. It's not something that runs under the hood; it's a dedicated field. Hopefully, there's a way of more or less quickly go through various sources like dotnet/fsharp, fslang-X repos, Sergeys blog and others, and then distill some concise articles.

**Reply by baronfel** [reply]
Good recent places to start on what's been happening would be the first and last talks of the recent fsharpconf, which can be found on the dotnet YouTube channel. The first is Kathleen and a few friends, the last is the F# team talking about features and stuff.

**Reply by SchlenkR** [reply]
The one with Kathleen, Vlad and you - yeah I got that in view. The other one I guess not, but I'll find. Thanks got the hint.

**Comment by SchlenkR** [comment]
I'm working on an idea and attempt for making F# language and compiler progression more prominent, currently by playing with web-scraping fslang-suggestions and dotnet/fsharp repos + manuel editing to render some small pages, giving an overview of the current state and their transitions. I thought that this might be possible with a minimal effort in order to keep such a thing alive, but still providing value to the viewer. It's a very rough idea + fiddling around. Roughly, it might come out to something like: A thing that could be "done" (in the past or fufute) is the last element of a stream of events, made up of various non-formally coupled sources.

@vzarytovskii or @cartermp:

- Since I'm not involved in the processes, could a statement be formulated like "Usually, in 80% of the cases, de-facto things work like ...". Then I'd be interested to understand how that could look like.
- Maybe: Simply by putting labels on issues / PRs, etc. might be very helpful to at least raise some triggers for getting focus on things, and to have a closer look. "Approved In Principle" is / was one of those labels. If there were mor of such soft conventions, it could help...

**An example:**

https://github.com/fsharp/fslang-suggestions/issues/1262

![grafik](https://github.com/fsharp/fslang-design/assets/472222/fbbf13bf-8e6c-4271-8325-7cda48d12a66)

Is there a way to understand if work started on a fslang-suggestion of a (bigger) feature in dotnet/fsharp? Just from the suggestions, it seems like it's suddenly done (I mean: That's an information - and a great one indeed :) )

Pls excuse if my language might sound a little bit "clumsy"; I tried my best. And thank you for your support.

**Reply by T-Gro** [reply]
The screenshot is actually a very good example.

As soon as a PR is **raised** (and that is very often a starting work in progress, just a draft), the original suggestion is being **linked**. The linkage is there while the PR is still open, independent of being later merged or not. This tells you that the work has been started.

**Reply by T-Gro** [reply]
The label "Merged" on the right side is automatically derived by GH. Before it was finished, the linkage was there as well, just the label would have been green and saying "Open".

**Comment by realvictorprm** [comment]
Side note, the title can be fixed to `state` instead of `sate` :smile:

**Reply by vzarytovskii** [reply]
I kinda liked the `sate`... :(
